### PR TITLE
Multithread improvements

### PR DESCRIPTION
**Pull request by pablogs9**
Required: https://github.com/eProsima/Micro-XRCE-DDS-Client/pull/289

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -22,7 +22,7 @@ jobs:
         - name: Download dependencies
           run: |
             git clone -b foxy https://github.com/eProsima/Micro-CDR src/Micro-CDR
-            git clone -b foxy https://github.com/eProsima/Micro-XRCE-DDS-Client src/Micro-XRCE-DDS-Client
+            git clone -b fix/session_ping https://github.com/eProsima/Micro-XRCE-DDS-Client src/Micro-XRCE-DDS-Client
```
Comment (pablogs9) [inline .github/workflows/ci.yml:25]:
```suggestion
            git clone -b ros2 https://github.com/eProsima/Micro-XRCE-DDS-Client src/Micro-XRCE-DDS-Client
```

Comment (pablogs9) [reply .github/workflows/ci.yml:25]: When https://github.com/eProsima/Micro-XRCE-DDS-Client/pull/289 merged

File: rmw_microxrcedds_c/src/rmw_wait.c [hunk 2]
```diff
@@ -34,7 +34,14 @@ rmw_wait(
   (void)events;
   (void)wait_set;
 
+  // With `rmw_uxrce_wait_mutex` member `need_to_be_ran` is protected.
+  // `session_memory` itself is not protected because it is not modified between
+  // rmw_init and rmw_shutdown, and rmw_wait cannot be called concurrently with
```
Comment (Acuadros95) [inline rmw_microxrcedds_c/src/rmw_wait.c:39]: Maybe we should add a multi-thread section to `Programming with rcl and rclc
` or XRCE-Client Documentation to add this guidelines.

Comment (pablogs9) [reply rmw_microxrcedds_c/src/rmw_wait.c:39]: Great idea

File: rmw_microxrcedds_c/src/callbacks.c [hunk 3]
```diff
@@ -136,13 +142,15 @@ void on_request(
       {
         put_memory(&static_buffer_memory, memory_node);
         return;
```
Comment (Acuadros95) [inline rmw_microxrcedds_c/src/callbacks.c:144]:
```suggestion
```

File: rmw_microxrcedds_c/src/callbacks.c [hunk 4]
```diff
@@ -173,6 +181,8 @@ void on_reply(
         custom_client, custom_client->qos);
       if (!memory_node) {
         RMW_SET_ERROR_MSG("Not available static buffer memory node");
+        UXR_UNLOCK(&static_buffer_memory.mutex);
```
Comment (Acuadros95) [inline rmw_microxrcedds_c/src/callbacks.c:184]: `UXR_LOCK(&static_buffer_memory.mutex);` missing on `on_reply` method

File: rmw_microxrcedds_c/src/callbacks.c [hunk 5]
```diff
@@ -186,13 +196,14 @@ void on_reply(
       {
         put_memory(&static_buffer_memory, memory_node);
         return;
```
Comment (Acuadros95) [inline rmw_microxrcedds_c/src/callbacks.c:198]:
```suggestion
```

File: rmw_microxrcedds_c/src/callbacks.c (hunk 4 above)
Comment (pablogs9) [reply rmw_microxrcedds_c/src/callbacks.c:185]: Solved

File: rmw_microxrcedds_c/src/rmw_microros/ping.c [hunk 6]
```diff
@@ -61,11 +60,13 @@ rmw_ret_t rmw_uros_ping_agent(
...
+    // There is a session available to ping. Using session.
     rmw_uxrce_mempool_item_t * item = session_memory.allocateditems;
     do {
       rmw_context_impl_t * context = (rmw_context_impl_t *)item->data;
 
-      success = uxr_ping_agent_attempts(&context->transport.comm, timeout_ms, attempts);
+      success = uxr_ping_agent_session(&context->session, timeout_ms, attempts);
+
       item = item->next;
     } while (NULL != item && !success);
```
Comment (Acuadros95) [inline rmw_microxrcedds_c/src/rmw_microros/ping.c:71]: I still dont like the ping multiple sessions approach.

Maybe we could change this to accept a `rmw_context_impl_t` as argument? Wich could be `NULL`.

Comment (pablogs9) [reply rmw_microxrcedds_c/src/rmw_microros/ping.c:71]: Don't understand, can you make a proposal?

**Review by Acuadros95 — APPROVED** [review]

**Comment by pablogs9** [comment]
@mergify backport foxy main

**Pull request merged**